_posts/2015-06-09-Fostering-Creativity-with-Semantically-Rich-APIs.md

```diff
@@ -0,0 +1,115 @@
+---
+title: Semantically Rich APIs
+excerpt: Fostering Creativity with Semantically Rich APIs
+layout: post
+date: 2015-06-09 00:00:00 +0200
+author: stephen
+published: true
+---
+
+In our recent blog post, Kyle Fuller [wrote about](http://blog.apiary.io/2015/06/04/Hyperdrive/) the benefits of decoupling semantics from implementation details for building better API clients. I'd like to explore how this decoupled approach also benefits the API design process—it's quite nice to decouple your mind from implementation details as well.
+
+## The Difficulty with Common Approaches
+
+Many modern approaches in designing and building APIs couple not only clients and servers to implementation details, but also couple thought, creativity, and shared understanding to that implementation. Coupling servers and clients is a problem that is easily seen, and in theory no one wants a coupled approach. Kyle Fuller wrote in his post:
+
+> Imagine entering an API without knowing about any implementation details, but instead just understanding the semantics of the API. We can get the implementation details of the API at run-time to allow us to change them at any point.
+
+In waiting until runtime to get these specific details, Hyperdrive allows for building clients that are highly evolvable and don't break when things are changed on the server.
+
+Coupling on the other hand hinders this API evolvability—among other things—but it also hinders API design and creativity for designers and consumers. I think this is something we miss. We're all happy to have won the battle over technologies like SOAP and WSDL, but we fail to see our tools and methodologies are leading us right back to where we were.
+
+## Chewing the CRUD
+
+If you survey the modern API space, you'll see this coupled design pattern is the norm, albeit it's a limited one. The first design step in this pattern is usually to carefully craft a URL that conveys to the users the meaning of the content. The next step is to pick the HTTP method the users will use in interacting with that URL. The thought, design, and creativity is directly coupled with the HTTP implementation in this process, and pushes the question "What does my API do?" to later steps in the design process.
+
+This approach maps directly into the concept of CRUD, which is frequently used to describe this kind of interface design. The term CRUD is from the area of database and encompasses the actions you take on persistent data. It stands for "Create, Read, Update, and Delete," which are common operations associated with interacting with persisted data. In the API world, these general semantics are matched up with HTTP methods, such as POST/PUT, GET, PUT/PATCH, and DELETE respectively. Using CRUD with HTTP APIs, a user performs these actions on a resource's URL.
```

Comment:
I feel it is important to explain the coupling, most people know what CRUD is. Explaining CRUD feels unnecessary to me as it does not contribute to the story (?)